core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,229 @@
+   * order, but also used if an error occurrs before {@code realStream} is set.
+   */
   private ClientStreamListener listener;
-  @GuardedBy("this")
   private ClientStream realStream;
+  /** The stream was cancelled before {@code realStream} was set. */
   @GuardedBy("this")
   private Status error;
-
-  @GuardedBy("this")
-  private final List<PendingMessage> pendingMessages = new LinkedList<PendingMessage>();
-  private boolean messageCompressionEnabled;
-  @GuardedBy("this")
-  private boolean pendingHalfClose;
-  @GuardedBy("this")
-  private int pendingFlowControlRequests;
-  @GuardedBy("this")
-  private boolean pendingFlush;
   @GuardedBy("this")
-  private Compressor compressor;
-  @GuardedBy("this")
-  private Decompressor decompressor;
-
-  static final class PendingMessage {
-    final InputStream message;
-    final boolean shouldBeCompressed;
+  private List<Runnable> pendingCalls = new ArrayList<Runnable>();
 
-    public PendingMessage(InputStream message, boolean shouldBeCompressed) {
-      this.message = message;
-      this.shouldBeCompressed = shouldBeCompressed;
+  /**
+   * Transfers all pending and future requests and mutations to the given stream.
+   *
+   * <p>No-op if either this method or {@link #cancel} have already been called.
+   */
+  final void setStream(ClientStream stream) {
+    synchronized (this) {
+      if (error != null || realStream != null) {
+        return;
+      }
+      realStream = checkNotNull(stream, "stream");
     }
-  }
 
-  @Override
-  public synchronized void setAuthority(String authority) {
-    checkState(listener == null, "must be called before start");
-    checkNotNull(authority, "authority");
-    if (realStream == null) {
-      this.authority = authority;
-    } else {
-      realStream.setAuthority(authority);
+    List<Runnable> toRun = new ArrayList<Runnable>();
+    while (true) {
+      synchronized (this) {
+        if (pendingCalls.isEmpty()) {
+          pendingCalls = null;
+          passThrough = true;
+          break;
+        }
+        // Since there were pendingCalls, we need to process them. To maintain ordering we can't set
+        // passThrough=true until we run all pendingCalls, but new Runnables may be added after we
+        // drop the lock. So we will have to re-check pendingCalls.
+        List<Runnable> tmp = toRun;
+        toRun = pendingCalls;
+        pendingCalls = tmp;
+      }
+      for (Runnable runnable : toRun) {
+        // Must not call transport while lock is held to prevent deadlocks.
+        // TODO(ejona): exception handling
+        runnable.run();
+      }
+      toRun.clear();
     }
   }
 
-  @Override
-  public void start(ClientStreamListener listener) {
+  /**
+   * Enqueue the runnable or execute it now. Call sites that may be called many times may want avoid
+   * this method if {@code passThrough == true}.
+   */
+  private void delayOrExecute(Runnable runnable) {
```

Comment:
Might be worth mentioning that this does not preserve runnable order if called from multiple threads